File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/activities/IntroActivity.java
package com.habitrpg.android.habitica.ui.activities;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v4.view.ViewPager;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;

import com.habitrpg.android.habitica.HabiticaBaseApplication;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.components.AppComponent;
import com.habitrpg.android.habitica.data.ApiClient;
import com.habitrpg.android.habitica.helpers.RxErrorHandler;
import com.habitrpg.android.habitica.ui.fragments.setup.IntroFragment;
import com.viewpagerindicator.IconPageIndicator;
import com.viewpagerindicator.IconPagerAdapter;

import javax.inject.Inject;

import butterknife.BindView;

public class IntroActivity extends BaseActivity implements View.OnClickListener, ViewPager.OnPageChangeListener {

    @Inject
    public ApiClient apiClient;
    @BindView(R.id.view_pager)
    ViewPager pager;
    @BindView(R.id.view_pager_indicator)
    IconPageIndicator indicator;
    @BindView(R.id.skipButton)
    Button skipButton;
    @BindView(R.id.finishButton)
    Button finishButton;

    @Override
    protected int getLayoutResId() {
        return R.layout.activity_intro;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        HabiticaBaseApplication.getComponent().inject(this);

        setupIntro();
        indicator.setViewPager(pager);

        this.skipButton.setOnClickListener(this);
        this.finishButton.setOnClickListener(this);

        apiClient.getContent()
                .subscribe(contentResult -> {}, RxErrorHandler.handleEmptyError());

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            Window window = getWindow();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                window.setStatusBarColor(ContextCompat.getColor(this, R.color.black_20_alpha));
            }
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS);
        }

    }

    @Override
    protected void injectActivity(AppComponent component) {
        component.inject(this);
    }

    private void setupIntro() {
        android.support.v4.app.FragmentManager fragmentManager = getSupportFragmentManager();

        pager.setAdapter(new PagerAdapter(fragmentManager));

        pager.addOnPageChangeListener(this);
    }

    @Override
    public void onClick(View v) {
        finishIntro();
    }

    private void finishIntro() {
        Intent intent = new Intent(this, LoginActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);

        this.startActivity(intent);
        overridePendingTransition(0, R.anim.activity_fade_out);
        finish();
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
    }

    @Override
    public void onPageSelected(int position) {
        if (position == 2) {
            this.finishButton.setVisibility(View.VISIBLE);
        } else {
            this.finishButton.setVisibility(View.GONE);
        }
    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }

    private class PagerAdapter extends FragmentPagerAdapter implements IconPagerAdapter {
        public PagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int position) {
            IntroFragment fragment = new IntroFragment();

            switch (position) {
                case 0: {
                    fragment.setImage(ResourcesCompat.getDrawable(getResources(), R.drawable.intro_1, null));
                    fragment.setSubtitle(getString(R.string.intro_1_subtitle));
                    fragment.setTitleImage(ResourcesCompat.getDrawable(getResources(), R.drawable.intro_1_title, null));
                    fragment.setDescription(getString(R.string.intro_1_description, getString(R.string.habitica_user_count)));
                    fragment.setBackgroundColor(ContextCompat.getColor(IntroActivity.this, R.color.brand_300));
                    break;
                }
                case 1: {
                    fragment.setImage(ResourcesCompat.getDrawable(getResources(), R.drawable.intro_2, null));
                    fragment.setSubtitle(getString(R.string.intro_2_subtitle));
                    fragment.setTitle(getString(R.string.intro_2_title));
                    fragment.setDescription(getString(R.string.intro_2_description));
                    fragment.setBackgroundColor(ContextCompat.getColor(IntroActivity.this, R.color.blue_10));
                    break;
                }
                case 2: {
                    fragment.setImage(ResourcesCompat.getDrawable(getResources(), R.drawable.intro_3, null));
                    fragment.setSubtitle(getString(R.string.intro_3_subtitle));
                    fragment.setTitle(getString(R.string.intro_3_title));
                    fragment.setDescription(getString(R.string.intro_3_description));
                    fragment.setBackgroundColor(ContextCompat.getColor(IntroActivity.this, R.color.red_100));
                    break;
                }
            }

            return fragment;
        }

        @Override
        public int getIconResId(int index) {
            return R.drawable.indicator_diamond;
        }

        @Override
        public int getCount() {
            return 3;
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/activities/IntroActivity.kt
package com.habitrpg.android.habitica.ui.activities

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.core.view.WindowCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayoutMediator
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.data.ContentRepository
import com.habitrpg.android.habitica.databinding.ActivityIntroBinding
import com.habitrpg.android.habitica.extensions.setNavigationBarDarkIcons
import com.habitrpg.android.habitica.ui.fragments.setup.IntroFragment
import com.habitrpg.common.habitica.helpers.ExceptionHandler
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class IntroActivity : BaseActivity() {
    private lateinit var binding: ActivityIntroBinding

    @Inject
    lateinit var contentRepository: ContentRepository

    override fun getLayoutResId(): Int {
        return R.layout.activity_intro
    }

    override fun getContentView(layoutResId: Int?): View {
        binding = ActivityIntroBinding.inflate(layoutInflater)
        return binding.root
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setupIntro()

        binding.skipButton.setOnClickListener { finishIntro() }
        binding.finishButton.setOnClickListener { finishIntro() }

        lifecycleScope.launch(ExceptionHandler.coroutine()) {
            contentRepository.retrieveContent()
        }
    }

    override fun onResume() {
        super.onResume()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            window.isNavigationBarContrastEnforced = false
            val controller = WindowCompat.getInsetsController(window, window.decorView)
            controller.isAppearanceLightNavigationBars = false
            controller.isAppearanceLightStatusBars = false
            window.setNavigationBarDarkIcons(false)
        }
    }

    private fun setupIntro() {
        setViewPagerAdapter()
        binding.viewPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                if (position == 2) {
                    binding.finishButton.visibility = View.VISIBLE
                } else {
                    binding.finishButton.visibility = View.GONE
                }
            }
        })
    }

    private fun finishIntro() {
        val intent = Intent(this, LoginActivity::class.java)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)

        this.startActivity(intent)
        overridePendingTransition(0, R.anim.activity_fade_out)
        finish()
    }

    private fun setViewPagerAdapter() {
        val fragmentManager = supportFragmentManager
        val viewPagerAdapter = object : FragmentStateAdapter(fragmentManager, lifecycle) {
            override fun createFragment(position: Int): Fragment {
                val fragment = IntroFragment()
                configureFragment(fragment, position)
                return fragment
            }

            override fun getItemCount(): Int {
                return 3
            }
        }
        binding.viewPager.adapter = viewPagerAdapter
        TabLayoutMediator(binding.viewPagerIndicator, binding.viewPager) { tab, position -> }.attach()
    }

    private fun configureFragment(
        fragment: IntroFragment,
        position: Int
    ) {
        when (position) {
            0 -> {
                fragment.setImage(ResourcesCompat.getDrawable(resources, R.drawable.intro_1, null))
                fragment.setSubtitle(getString(R.string.intro_1_subtitle))
                fragment.setTitleImage(
                    ResourcesCompat.getDrawable(
                        resources,
                        R.drawable.intro_1_title,
                        null
                    )
                )
                fragment.setDescription(
                    getString(
                        R.string.intro_1_description,
                        getString(R.string.habitica_user_count)
                    )
                )
                fragment.setBackgroundColor(
                    ContextCompat.getColor(
                        this@IntroActivity,
                        R.color.brand_300
                    )
                )
            }

            1 -> {
                fragment.setImage(ResourcesCompat.getDrawable(resources, R.drawable.intro_2, null))
                fragment.setSubtitle(getString(R.string.intro_2_subtitle))
                fragment.setTitle(getString(R.string.intro_2_title))
                fragment.setDescription(getString(R.string.intro_2_description))
                fragment.setBackgroundColor(
                    ContextCompat.getColor(
                        this@IntroActivity,
                        R.color.blue_10
                    )
                )
            }

            2 -> {
                fragment.setImage(ResourcesCompat.getDrawable(resources, R.drawable.intro_3, null))
                fragment.setSubtitle(getString(R.string.intro_3_subtitle))
                fragment.setTitle(getString(R.string.intro_3_title))
                fragment.setDescription(getString(R.string.intro_3_description))
                fragment.setBackgroundColor(
                    ContextCompat.getColor(
                        this@IntroActivity,
                        R.color.red_100
                    )
                )
            }
        }
    }
}
